src/Misc/coroutine.ts

```diff
@@ -0,0 +1,133 @@
+                    resume();
+                }
+            },
+            (error: any) => {
+                // If the coroutine threw an error, report the error.
+                onError(error);
+            });
+    }
+
+    resume();
+}
+
+// This is a helper type to extract the return type of a Coroutine<T>. It is conceptually very similar to the Awaited<T> utility type.
+type ExtractCoroutineReturnType<T> = T extends Coroutine<infer TReturn> ? TReturn : never;
+
+/**
+ * Given a function that returns a Coroutine<T>, produce a function with the same parameters that returns a T.
+ * The returned function runs the coroutine synchronously.
+ * @param coroutineFactory A function that returns a Coroutine<T>.
+ * @returns A function that runs the coroutine synchronously.
+ */
+export function makeSyncFunction<TReturn, TCoroutineFactory extends (...params: any[]) => Coroutine<TReturn>>(coroutineFactory: TCoroutineFactory): (...params: Parameters<TCoroutineFactory>) => ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>> {
+    return (...params: Parameters<TCoroutineFactory>): ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>> => {
+        // Create the coroutine from the factory function.
+        const coroutine = coroutineFactory(...params);
+
+        // Run the coroutine with the inline scheduler, storing the returned value, or re-throwing the error (since the error callback will be called synchronously by the inline scheduler).
+        let result: TReturn | undefined;
+        runCoroutine(coroutine, inlineScheduler, (r: TReturn) => result = r, (e: any) => { throw e; });
+
+        // Synchronously return the result of the coroutine.
+        return result as ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>>; // TODO: How can we remove this cast?
+    };
+}
+
+/**
+ * Given a function that returns a Coroutine<T>, product a function with the same parameters that returns a Promise<T>.
+ * The returned function runs the coroutine asynchronously, yield control of the execution context occasionally to enable a more responsive experience.
+ * @param coroutineFactory A function that returns a Coroutine<T>.
+ * @returns A function that runs the coroutine asynchronously.
+ */
+export function makeAsyncFunction<TReturn, TCoroutineFactory extends (...params: any[]) => Coroutine<TReturn>>(coroutineFactory: TCoroutineFactory): (...params: Parameters<TCoroutineFactory>) => Promise<ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>>> {
+    return (...params: Parameters<TCoroutineFactory>): Promise<ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>>> => {
+        // Create the coroutine from the factory function.
+        const coroutine = coroutineFactory(...params);
+
+        // Run the coroutine with a yielding scheduler, resolving or rejecting the result promise when the coroutine finishes.
+        const result = new Deferred<TReturn>();
```

Comment:
Why not using a regular promise here ?

Reply:
I think I needed it in an older iteration of the code, but clearly that's no longer the case. Switched to Promise.